### PR TITLE
Run remove-virtual-functions on a per-function basis

### PR DESCRIPTION
**Pull request by smowton**
This introduces a mechanism to re-number a single GOTO program, since various parts of the CBMC codebase will assume the globally-unique-location-numbers invariant holds and it's safest to ensure it holds as much of the time as possible, then uses it to move remove-virtual-functions to run as each function is elaborated, rather than all in one go at the end.

### CODE REVIEW

File: src/goto-programs/goto_functions_template.h
Comment (tautschnig) [inline src/goto-programs/goto_functions_template.h]: Is that `-` a minus (subtraction) or (more likely) indicating a range? Would you mind being clinical here and use a notation that unambiguously states whether that's a (half) open interval?

File: src/goto-programs/goto_functions_template.h
Comment (tautschnig) [inline src/goto-programs/goto_functions_template.h]: I don't think this should be `public`?

File: src/goto-programs/goto_model.h
Comment (tautschnig) [inline src/goto-programs/goto_model.h]: What exactly is the value of this?

File: src/goto-programs/goto_model.h
Comment (tautschnig) [inline src/goto-programs/goto_model.h]: Wrong commit. (Though in my earlier comment it doubted the overall value of this interface as well.)

File: src/goto-programs/lazy_goto_model.cpp
Comment (tautschnig) [inline src/goto-programs/lazy_goto_model.cpp]: Isn't this just a container for `goto_modelt`? I find this pretty weird.

**Comment by smowton** [comment]
@tautschnig all comments addressed, except for the interface, whose purpose is to permit GOTO program renumbering (a non-const operation on `goto_functionst`) without providing a non-const reference to the entire function map, which should explicitly **not** be perturbed while post-processing a single function in isolation.

Alternatively solutions:
* Rewrite `location_number` as `union { struct { uint32_t function_number; uint32_t instruction_number; }; uint64_t location_number; }`, thus permitting GOTO program re-numbering without coordinating via `goto_functionst` again (good thing to do, more churny)
* Make `unused_location_number` mutable, so a const ref to `goto_functionst` suffices (don't like it; mutable is rather a blunt instrument compared to supplying a specific interface that exposes the required feature)
* Maintain a back-pointer from `goto_programt` to the function map it is in (error-prone; can a particular program (perhaps briefly) be in more than one map, for example?)

**Comment by smowton** [comment]
@NathanJPhillips: please review (can't tag you due to your lack of write access to cbmc)

**Comment by tautschnig** [comment]
> [...] purpose is to permit GOTO program renumbering (a non-const operation on `goto_functionst`) without providing a non-const reference to the entire function map, which should explicitly **not** be perturbed while post-processing a single function in isolation.

This sounds good to me, and certainly I wouldn't have complained at all had I read this comment in the source code. Would you mind adding it? ("Interface for renumbering the locations in a goto_programt" lacks the "why" that the above explanation does provide.)

**Review by tautschnig — APPROVED** [review]
Approving with the assumption that comments will be clarified (well, indeed they might just need to be shifted from the point-of-use to the definition).

**Comment by smowton** [comment]
Added a comment as requested.

File: src/goto-programs/goto_model.h
Comment (peterschrammel) [inline src/goto-programs/goto_model.h]: Nit pick: prefer protected members at bottom (as typical in the codebase).

File: src/goto-programs/goto_program_template.h
Comment (peterschrammel) [inline src/goto-programs/goto_program_template.h]: use `std::numeric_limits::max`

File: unit/testing-utils/load_java_class.cpp
Comment (peterschrammel) [inline unit/testing-utils/load_java_class.cpp]: It feels quite odd nonetheless that we introduced `goto_modelt` as a container for goto_functions and symbol_table to simplify our interfaces and now here we are creating interfaces that require even more parameters. What about wrapping a goto_function and its goto_model into a `goto_function_modelt`? Then you can call renumber() on the goto_function_model without exposing the underlying goto_model.

**Comment by smowton** [comment]
@peterschrammel you're quite right, I created `goto_model_functiont` which provides similar functionality but is much nicer. @tautschnig might want to look again since this has changed a bit.

File: src/goto-programs/goto_model.h
Comment (tautschnig) [inline src/goto-programs/goto_model.h:94]: Nit pick: "program order" may be ambiguous (or possibly wrong) here; the numbers are strictly increasing as they are laid out in the underlying storage (source text or a list of statements), which need not coincide with the order of execution.

Comment (smowton) [reply src/goto-programs/goto_model.h:94]: This is the wording used elsewhere in `goto-programs/goto_program_template.h` so I'll keep this as-is

**Comment by smowton** [comment]
Test-gen bump here: https://github.com/diffblue/test-gen/pull/1361

**Review by peterschrammel — APPROVED** [review]
This looks more elegant now.

**Pull request merged**
